### PR TITLE
Added --check flag for checking dependencies version without actually updating the Cargo.toml file.

### PR DESCRIPTION
**Pull request by 0xsouravm**
This PR introduces a new --check flag to the CLI tool, enabling users to verify if their project's dependencies are up-to-date with the specified versions without making any changes to the Cargo.toml file.

**Key Features:**

1. _Dependency Verification:_ With the --check flag, users can now verify the versions of all dependencies listed in Cargo.toml against the specified versions.

2. _Non-Intrusive:_ This feature reads and checks the versions without modifying the Cargo.toml file, making it a safe operation for continuous integration checks or pre-commit hooks.

### CODE REVIEW

File: src/main.rs
Comment (shawntabrizi) [inline src/main.rs]: shouldn't check input a version too? else which version are you checking against?

Comment (0xsouravm) [reply src/main.rs]: --check does take a version. The `required_unless_present = "version"` was put in place because you don't need to use the --version flag with the --check flag.

`psvm --check <VERSION>` works.

You don't need to do `psvm --check --version <VERSION>`

Comment (shawntabrizi) [reply src/main.rs]: ah. i would guess the syntax should be:

```
psvm --version <VERSION> --check
```

So you add check on top of the normal syntax.

Comment (shawntabrizi) [reply src/main.rs]: This is the pattern other tools use, like rustfmt:

https://github.com/rust-lang/rustfmt?tab=readme-ov-file#verifying-code-is-formatted

Comment (0xsouravm) [reply src/main.rs:56]: Oh alright got it. Updating the syntax

Comment (0xsouravm) [reply src/main.rs]: Done @shawntabrizi

File: src/check_deps.rs
Comment (Szegoo) [inline src/check_deps.rs]: To get rid of the warning since we are not reading the value until we assign it.

```suggestion
    let mut local_toml_version: &str;
```

**Review by Szegoo — APPROVED** [review]
Tested it locally and it works as expected. I left a few remarks to resolve warnings, but other than that the code looks overall good to me. (there is some code duplication that could perhaps be improved, but it should be fine as is)

File: src/check_deps.rs
Comment (Szegoo) [inline src/check_deps.rs]: nit

```suggestion
    check_dependencies_impl(cargo_toml_path, crates_versions, overwrite)?;
```

File: src/check_deps.rs
Comment (Szegoo) [inline src/check_deps.rs]: To resolve clippy warning

```suggestion
                return Err("Invalid Dependency Format".to_string().into());
```

Comment (Szegoo) [inline src/check_deps.rs:74]: To resolve clippy warning

```suggestion
                return Err("Invalid Dependency Format".to_string().into());
```

File: src/check_deps.rs
Comment (0xsouravm) [reply src/check_deps.rs]: I fixed the warnings you mentioned, should be good now. If possible can you point out the code duplications? I took the liberty in changing the original code structure to prevent some code duplication and modularise it. But if there's still some duplication, I would love to finish it up before this is merged.

Comment (Szegoo) [reply src/check_deps.rs]: This is probably the approach I would have taken. This way, we would only need to add a new `check_dependency` function where your checking logic would be placed.

NOTE: this is a diff against the main branch.

```diff
diff --git a/src/main.rs b/src/main.rs
index b0dd6b0..79cbaa3 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -49,6 +49,10 @@ struct Command {
     /// List available versions.
     #[clap(short, long)]
     list: bool,
+
+    /// Checks if deps have the same version as the specified Polkadot SDK version.
+    #[clap(short, long)]
+    check: bool,
 }
 
 #[tokio::main]
@@ -73,7 +77,12 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
     let crates_versions: BTreeMap<String, String> =
         get_version_mapping_with_fallback(DEFAULT_GIT_SERVER, &version).await?;
 
-    update_dependencies(&cargo_toml_path, &crates_versions, cmd.overwrite)?;
+    check_dependencies(
+        &cargo_toml_path,
+        &crates_versions,
+        cmd.overwrite,
+        !cmd.check,
+    )?;
 
     Ok(())
 }
@@ -94,12 +103,13 @@ fn validate_workspace_path(mut path: PathBuf) -> Result<PathBuf, Box<dyn std::er
     Ok(path)
 }
 
-fn update_dependencies(
+fn check_dependencies(
     cargo_toml_path: &Path,
     crates_versions: &BTreeMap<String, String>,
     overwrite: bool,
+    update: bool,
 ) -> Result<(), Box<dyn std::error::Error>> {
-    let cargo_toml = update_dependencies_impl(cargo_toml_path, crates_versions, overwrite)?;
+    let cargo_toml = check_dependencies_impl(cargo_toml_path, crates_versions, overwrite, update)?;
 
     match cargo_toml {
         Some(new_content) => {
@@ -117,10 +127,11 @@ fn update_dependencies(
     Ok(())
 }
 
-fn update_dependencies_impl(
+fn check_dependencies_impl(
     cargo_toml_path: &Path,
     crates_versions: &BTreeMap<String, String>,
     overwrite: bool,
+    update: bool,
 ) -> Result<Option<String>, Box<dyn std::error::Error>> {
     let cargo_toml_content = fs::read_to_string(cargo_toml_path)?;
     let mut cargo_toml: DocumentMut = cargo_toml_content.parse()?;
@@ -132,7 +143,7 @@ fn update_dependencies_impl(
 
     for table in ["dependencies", "dev-dependencies", "build-dependencies"].iter() {
         if let Some(toml_edit::Item::Table(dep_table)) = deps.get_mut(table) {
-            update_table_dependencies(dep_table, crates_versions, overwrite);
+            check_table_dependencies(dep_table, crates_versions, overwrite, update);
         }
     }
 
@@ -144,10 +155,11 @@ fn update_dependencies_impl(
     }
 }
 
-pub fn update_table_dependencies(
+pub fn check_table_dependencies(
     dep_table: &mut toml_edit::Table,
     crates_versions: &BTreeMap<String, String>,
     overwrite: bool,
+    update: bool,
 ) {
     for (dep_key, dep_value) in dep_table.iter_mut() {
         let dep_key_str = dep_key.get();
@@ -167,42 +179,55 @@ pub fn update_table_dependencies(
             continue;
         };
 
-        if let Some(table) = dep_value.as_table_like_mut() {
-            if !overwrite && table.get("path").is_some() {
-                continue;
-            }
+        if update {
+            update_dependency(dep_value, dep_key_str, crate_version, overwrite);
+        } else {
+            // TODO: check
+        }
+    }
+}
 
-            table.remove("git");
-            table.remove("rev");
-            table.remove("branch");
-            table.remove("tag");
-            table.remove("path");
+fn update_dependency(
+    dep_value: &mut toml_edit::Item,
+    dep_key_str: &str,
+    crate_version: &String,
+    overwrite: bool,
+) {
+    if let Some(table) = dep_value.as_table_like_mut() {
+        if !overwrite && table.get("path").is_some() {
+            return;
+        }
 
-            let mut new_table = toml_edit::InlineTable::default();
+        table.remove("git");
+        table.remove("rev");
+        table.remove("branch");
+        table.remove("tag");
+        table.remove("path");
 
-            // Directly create a `toml_edit::Value` for the version
-            new_table.get_or_insert(
-                "version",
-                toml_edit::value(crate_version.clone()).as_value().unwrap(),
-            );
+        let mut new_table = toml_edit::InlineTable::default();
 
-            for (key, value) in table.iter() {
-                // Ensure we're inserting `Value`s, not `Item`s
-                if key != "version" && value.is_value() {
-                    new_table.get_or_insert(key, value.as_value().unwrap().clone());
-                }
-            }
-            new_table.fmt();
+        // Directly create a `toml_edit::Value` for the version
+        new_table.get_or_insert(
+            "version",
+            toml_edit::value(crate_version.clone()).as_value().unwrap(),
+        );
 
-            // Replace the original table-like item with the new inline table
-            *dep_value = toml_edit::Item::Value(toml_edit::Value::InlineTable(new_table));
-        } else if dep_value.is_str() {
-            *dep_value = toml_edit::value(crate_version.clone());
-        } else {
-            log::error!("Unexpected dependency value type for {}", dep_key_str);
-            continue;
+        for (key, value) in table.iter() {
+            // Ensure we're inserting `Value`s, not `Item`s
+            if key != "version" && value.is_value() {
+                new_table.get_or_insert(key, value.as_value().unwrap().clone());
+            }
         }
+        new_table.fmt();
 
-        log::debug!("Setting {} to {}", dep_key_str, crate_version);
+        // Return the new inline table
+        *dep_value = toml_edit::Item::Value(toml_edit::Value::InlineTable(new_table))
+    } else if dep_value.is_str() {
+        *dep_value = toml_edit::value(crate_version.clone());
+    } else {
+        log::error!("Unexpected dependency value type for {}", dep_key_str);
+        return;
     }
+
+    log::debug!("Setting {} to {}", dep_key_str, crate_version);
 }
```

Comment (0xsouravm) [reply src/check_deps.rs]: Understood. I initially thought to merge the functions and keep the original `update_dependencies` function and add check_dependency logic to it. But it kinda felt weird to both update and check dependencies by calling the same function which was either called `update_dependencies(...)` or `check_dependencies(...)`. It is not intuitive as to what the function does because it both updated and checked. That's why I decided to move with separate functions for each of the tasks.

Comment (Szegoo) [reply src/check_deps.rs]: I am not saying this is definitely the 'right approach'.

@shawntabrizi what are your thoughts?

File: src/versions.rs
Comment (ggwpez) [inline src/versions.rs]: Revert please to keep the diff small.

**Review by ggwpez — CHANGES_REQUESTED** [review]
The MR description fails to mention that this refactors a lot of things without it being the objective.

Please just add the minimal changes to achieve the goal here. Just keep the TOML files in memory and do all the updates but dont write them back to disk if there is a `-c` flag (and instead error if there were changes).

File: src/versions.rs
Comment (0xsouravm) [reply src/versions.rs]: Do I still show the logs for which dependencies need a version upgrade or do I just exit saying there are dependencies that need update?

Comment (ggwpez) [reply src/versions.rs]: If the tool already prints that for free then yea it can be helpful. But adding it additionally for now is not needed i think.

Comment (0xsouravm) [reply src/versions.rs]: I reverted back and made minimal changes @ggwpez

File: src/versions.rs
Comment (ggwpez) [inline src/versions.rs:1]: I think this file had no logic changes? Then please undo until we have a formatter in the CI - otherwise devs will change this back and forth depending on their exact local Rust version.

**Review by ggwpez — APPROVED** [review]
LGTM, thanks!

File: src/versions.rs
Comment (0xsouravm) [reply src/versions.rs:1]: Done! @ggwpez

File: src/check_deps.rs
Comment (shawntabrizi) [reply src/check_deps.rs]: `let _ =` is an antipattern! Don't ever write code like that imo

**Review by shawntabrizi — APPROVED** [review]
Yeah, this is looking way better after updates and minification of the PR.

File: src/versions.rs
Comment (shawntabrizi) [reply src/versions.rs:1]: great update @0xsouravm

**Comment by shawntabrizi** [comment]
@ggwpez ill let you merge :)

File: src/versions.rs
Comment (0xsouravm) [reply src/versions.rs:1]: Thanks!

File: src/main.rs
Comment (Szegoo) [inline src/main.rs]:
```suggestion
            Ok(Some(new_content))
```

File: src/main.rs
Comment (Szegoo) [inline src/main.rs]:
```suggestion
            Err("Dependencies are not up to date".into())
```

Comment (shawntabrizi) [reply src/main.rs]: @0xsouravm probably some of these things get caught if you do `cargo clippy` and `cargo fmt`

I am guessing clippy would catch this kind of suggestion

File: src/main.rs
Comment (Szegoo) [inline src/main.rs]: Also, maybe we could rename this to `check_only` since the function is already called `update_dependencies`, so it is kind of implied that it would update the dependencies.

File: src/main.rs
Comment (Szegoo) [reply src/main.rs]: Yes

Comment (0xsouravm) [reply src/main.rs]: Sorry about that! It's done now @shawntabrizi @Szegoo

File: src/main.rs
Comment (0xsouravm) [reply src/main.rs]: Updated the flag name!

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/sponsorship-for-sub0-thailand-2024/10093/26

**Pull request merged**
